enhancements/image-references.md

```diff
@@ -0,0 +1,234 @@
+process of building a `FROM scratch` image layer is also identical.
+
+Today, some prior art exists in the OSBS, IIB and CVP workflows for building
+operator bundles. As we improve the Operator SDK tooling to create a straghtforward
+process for creating bundle images, we must make sure a seamless migration is possible.
+
+### Risks and Mitigations
+
+It will be critical that the design be vetted by all of the concerned parties, from
+operator manifest authors to CI system authors and productized pipeline authors to
+ensure that the UX is appropriate in all cases. Furthermore, the largest risk in the
+implementation here is not prioritizing a clean migration pathway for all current
+users who create bundle images, which would lead to further fragmentation of the
+ecosystem, which is directly opposed to the goal of this enhancement.
+
+## Design Details
+
+The definition of a minimally-viable operator bundle image will be changed to
+ensure that all image references in the contained manifests have been resolved 
+to a digest and had the pull specifications rewritten to refer to those digests.
+
+The only acceptable process for creating an operator bundle will be to run the
+`operator-sdk generate bundle` CLI, providing the manifests, metadata and image
+sources as input to the creation process.
+
+### Proposed UX
+
+A set of new files in the `metadata/` directory of a bundle will be authored by
+the manifest authors and build systems to record image replacement intent and
+execution.
+
+Operator manifest authors write a manifest that refers to images using some
+opaque string, and provide an `image-references.yaml` file alongside their
+manifests that declares which strings inside of their manifest are referring
+to pull specifications of images and names each occurence.
+
+The `metadata/image-references.yaml` file holds data in the following format, mapping
+common names of container images to their string placeholders in the manifests:
+
+```yaml
+imageReferences:
+- name: common-name
+  substitute: registry.svc.ci.openshift.org/openshift:image
+```
+
+This mapping, therefore, defines what needs to be replaced in manifests when
+they are bundled and identifies each replacement with a name. When
+`operator-sdk generate bundle` runs, it will require as input a second mapping
+from those names to literal image pull specifications and will run the replacement.
+In this manner, the configuration provided by the manifest author remains static
+regardless of the eventual replacement that a build system will execute.
+
+The second mapping required at build-time, known as `image-replacements.yaml`, will take
+a similar form, mapping the common names of images to their explicitly resolved pull
+specifications. This file will have the following format:
+
+```yaml
+imagePullSpecs:
+- name: common-name
+  pullSpec: quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:298d0b496f67a0ec97a37c6e930d76db9ae69ee6838570b0cd0c688f07f53780
```

Comment:
In most current cases, the maintainer provides a pull spec to a tag which the build system resolves to a digest. Having the build system do the digest pinning greatly improves the user experience. 

I'm wondering how the build system could make the jump from `registry.svc.ci.openshift.org/openshift:image` to `quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:298d0b496f67a0ec97a37c6e930d76db9ae69ee6838570b0cd0c688f07f53780`. It seems like we need an intermediate pullspec, e.g. `quay.io/openshift-release-dev/ocp-v4.0-art-dev:latest`.

One possibility is for maintainers, the ones using the build system, to provide a non-digest pullspec in `image-replacements.yaml`. Then, the build system could simply resolve the unresolved pullspecs in it at build time.

Does that make sense?

Reply:
I don't think there is any jump required here - based on how the `ImageStream` in `registry.svc.ci.openshift.org` is configured, the normal mechanism for determining image digests will resolve the layers either at the underlying registry - quay, in this case - or the OpenShift registry as configured. 

I *do* expect maintainers of the build system to change the non-digest pullspec in `image-replacements.yaml` to point to the "right thing" - the build system should know, given a semantic name, where it should live. For instance, CI will know that `my-operator` semantically means a tag in an internal registry. OSBS will know that `my-operator` is owned by a specific `dist-git` repo, etc (the config for this is already something you ask users to provide IIRC) and will know whether the build should use e.g. the staging or prod reigstry. The build system would set values as necessary and the build would resolve them to a digest by simply following the normal rules and expecting the value being resolved to be authoritative.

Reply:
`ImageStream` is not listed anywhere in this doc. Did you mean `ImageReference`, `ImagePullSpec`, or something else?

I think we're on the same page regarding how `image-replacements.yaml` will be used by a build system. The maintainer creates it to point to some agreed upon location, e.g. a floating tag in the production registry, then the build system is responsible for altering the pull specs in `image-replacements.yaml`, "on the fly" as part of the build process, to use a digest for example.

Reply:
`ImageStream` being the underlying technology that you end up using when you hit that particular registry, using the registry API, IIRC. Meaning, it's up to the admin who sets up the registry and not something the user cares about here.